upf/upf_pfcp_api.c

```diff
@@ -803,16 +808,140 @@ lookup_nwi (u8 * name)
   return pool_elt_at_index (gtm->nwis, p[0]);
 }
 
+static u32
+generate_teid (u32 start, u32 mask)
+{
+  return random_u32(&start) & (UINT32_MAX ^ mask);
+}
+
+static int
+teid_v4_lookup (upf_main_t *gtm, u32 teid, upf_upip_res_t *res)
+{
+  clib_bihash_kv_8_8_t kv, value;
+  gtpu4_tunnel_key_t key4;
+
+  key4.dst = res->ip4.as_u32;
+  key4.teid = teid;
+
+  kv.key = key4.as_u64;
+
+  if (PREDICT_FALSE
+	(clib_bihash_search_8_8 (&gtm->v4_tunnel_by_key, &kv, &value)))
+            return UPF_GTPU_ERROR_NO_SUCH_TUNNEL;
+
+  return 0;
+}
+
+
+static int
+teid_v6_lookup (upf_main_t *gtm, u32 teid, upf_upip_res_t *res)
+{
+  clib_bihash_kv_24_8_t kv, value;
+
+  kv.key[0] = res->ip6.as_u64[0];
+  kv.key[1] = res->ip6.as_u64[1];
+  kv.key[2] = teid;
+
+  if (PREDICT_FALSE
+	(clib_bihash_search_24_8 (&gtm->v6_tunnel_by_key, &kv, &value)))
+            return UPF_GTPU_ERROR_NO_SUCH_TUNNEL;
+
+  return 0;
+}
+
+
+static void
+teid_add_del_by_choose_id (upf_session_t *sx, u8 choose_id, u32 teid, u8 add)
+{
+  if (add)
+    {
+      sparse_vec_validate (sx->teid_by_choose_id_table, choose_id)[0] = teid;
+    }
+  else
+    {
+      sparse_vec_validate (sx->teid_by_choose_id_table, choose_id)[0] = 0;
+    }
+/*  if (add)
+  {
+    hash_set (sx->teid_by_choose_id_table, choose_id, teid);
+  } else
+  {
+    hash_unset(sx->teid_by_choose_id_table, teid);
+  }
+*/
+}
```

Comment:
this is a very non VPP way to use this
```suggestion
{
  u32 *n;

  if (add)
    {
      n = sparse_vec_validate (sx->teid_by_choose_id_table, choose_id);
      n[0] = teid;
    }
  else
    {
      n = sparse_vec_validate (sx->teid_by_choose_id_table, choose_id);
      n[0] = 0;
    }
}
```